### PR TITLE
Add GitProcess.md and Contributing.md

### PR DESCRIPTION
**Pull request by nfrisby**
`Contributing.md` is the entrypoint to the docs for new team members and community members.

### CODE REVIEW

File: ouroboros-consensus/docs/Contributing.md [hunk 1]
```diff
@@ -1,5 +1,35 @@
...
-See the
-https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/StyleGuide.md
-document.
+See the follow documents.
+
+  * this repository's [README][repo-README]; it links to design documents
+  * Consensus [style guide][consensus-style]
+  * Consensus [git process][consensus-git]
+  * Consensus [onboarding][consensus-onboarding]
+  * TODO quick-start guide
```
Comment (nfrisby) [inline ouroboros-consensus/docs/Contributing.md:9]: @EncodePanda I'm envisioning your PR #2885 document here. Do you think that's a good fit?

**Comment by nfrisby** [comment]
@coot -- please let me know if you'd like me to include more information here, since we're still as of yet sharing this repository between Network and Consensus.

File: ouroboros-consensus/docs/Contributing.md (hunk 1 above)
Comment (EncodePanda) [reply ouroboros-consensus/docs/Contributing.md:9]: yeah, I think so, let's get this PR in first, and then I will apply my changes accordingly

File: ouroboros-consensus/docs/Contributing.md [hunk 2]
```diff
@@ -1,5 +1,35 @@
-TODO eg git history goals etc
+# How to contribute to the Consensus Layer
 
-See the
-https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/StyleGuide.md
-document.
+See the follow documents.
+
+  * this repository's [README][repo-README]; it links to design documents
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Contributing.md:5]: To be frank, we have so MANY rules upfront that new contributor has to adhere to, it's a bit overwhelming for new comers.
It's not a comment to this PR, it's a more general observation.

File: ouroboros-consensus/docs/Contributing.md [hunk 3]
```diff
@@ -1,5 +1,35 @@
...
+  * Consensus [style guide][consensus-style]
+  * Consensus [git process][consensus-git]
+  * Consensus [onboarding][consensus-onboarding]
+  * TODO quick-start guide
+
+Our primary neighbors are the upstream [cardano-ledger-specs] and the downstream
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Contributing.md:11]: this line is not needed in `Contributing.md` - people who read it most likely are aware of that fact. It brings unnecessary noise IMHO.

File: ouroboros-consensus/docs/Contributing.md [hunk 4]
```diff
@@ -1,5 +1,35 @@
...
+[cardano-node].
+
+If you're not an IOHK employee, please fork the repository on GitHub and open
+PRs that way. We're excited about your involvement; thank you!
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Contributing.md:15]: I think this paragraph should be moved to the top of the file. Encourage the contribution, ground general expectations to the soon-to-be contributor.

File: ouroboros-consensus/docs/Contributing.md [hunk 5]
```diff
@@ -1,5 +1,35 @@
...
+
+If you're an IOHK employee, work with your manager and DevOps to acquire
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Contributing.md:17]: I think this should be moved to a separate file (IOHK.md) and only referenced here

File: ouroboros-consensus/docs/Contributing.md [hunk 6]
```diff
@@ -1,5 +1,35 @@
...
+permissions to push branches to and open PRs in this repository. If you're
+working as part of the Consensus Team, also acquire permissions to Approve PRs
+(in particular, our `bors` bot must accept your commands). You'll eventually
+also need to be in the `CODEOWNERS` file. The general IOHK onboarding should
+have discussed establishing hardware-based Two Factor Authentication on your
+GitHub account; that's another prerequisite.
+
+[repo-README]: https://github.com/input-output-hk/ouroboros-network/README.md
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Contributing.md:25]: Maybe we should not build cycles? I think README.md should be an entry point to the rest of the documentation, we should point from README to Contributing, not the other way around.

File: ouroboros-consensus/docs/Contributing.md [hunk 7]
```diff
@@ -1,5 +1,35 @@
...
+
+[consensus-style]: https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/StyleGuide.md
+
+[consensus-git]: https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/GitProcess.md
+
+[consensus-onboarding]: https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/Onboarding.md
+
+[cardano-ledger-specs]: https://github.com/input-output-hk/cardano-ledger-specs
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Contributing.md:33]: Those two links below should have separate chapters and some description like down/up stream dependency.

File: ouroboros-consensus/docs/GitProcess.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+and pass the tests that are enabled in CI. In particular, this process ensures
+`git bisect` can be as useful as possible -- when you reach for that kind of
+sledgehammer, you want it to work well.
+
+Note that CI enables `-Werror`, but we don't require you to enable it locally.
+If you'd like to opt-in to that or similar, add a `cabal.project.local` file.
+
+## Guidelines
+
+  * Write useful commit messages. There are lots of guides on the Internet about
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/GitProcess.md:37]: It's not a scope of this PR but maybe it is worth exploring some form of standardization. I found this blog post very useful years back https://chris.beams.io/posts/git-commit/

File: ouroboros-consensus/docs/GitProcess.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+motivation. They help reviewers do their job, future developers understand the
+past design process, troubleshooters debug changes in behavior, downstream
+developers understand recent changes and their context, and so on. In these kind
+of scenarios, commit messages are often the easiest documentation to read first.
+
+_Small and organized commits_: The commits themselves, beyond just their
+messages, also contribute to the story in their own way. Prefer small commits
+that include only necessary changes, and choose a commit order that seems most
+natural to you. Doing so often also makes it easier to write useful commit
+messages. Rebase before merging, so the repository's full story is as linear as
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/GitProcess.md:19]: Have we ever considered to do a rebase and then fast-forward merge to origin/master (to achieve full linearizability)? General question, not actual comment to this PR

File: ouroboros-consensus/docs/GitProcess.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    for final review and merging. I often open my PR as Draft PR in order to get
+    early feedback and then later promote it to a proper PR. Or if I realize
+    there is big hole in my PR, I will convert it back into a Draft PR.
+
+    This is far superior to including "WIP" or "DO NOT MERGE" in the PR title,
+    or a WIP label, etc.
+
+    [github-draft-pr]: https://github.blog/2019-02-14-introducing-draft-pull-requests
+
+  * We're currently bridging the gap with both GitHub Issues and tickets in
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/GitProcess.md:74]: This is IOHK specific. Do we want to keep this in this document? We ad that discussion the other day that we want to keep those docs as much IOHK agnostic as we possibly can. 
Maybe some additional comment here like "for issues that originated from IOHK's JIRA please ...."

File: ouroboros-consensus/docs/GitProcess.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    IOHK's JIRA. For now, we are therefore also mentioning JIRA Issues such as
+    `CAD-NNNN` in PR titles and PR descriptions.
+
+  * Your branch name is impermanent, so it's less important than the above.
+    However, we prefer the format `username/issue-NNNN-short-description`.
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/GitProcess.md:79]: Not quite clear here on the intent. Is the `NNNN` from the JIRA's `CAD-NNNN` or the github issue's `NNNN`?

File: ouroboros-consensus/docs/Onboarding.md [hunk 12]
```diff
@@ -146,6 +146,4 @@ The following artifacts influence and/or describe the Consensus implementation.
 
 See
 https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/Contributing.md
-and as usual the
-https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/README.md
-file.
+for all of our advice about approaching this repository.
```
Comment (EncodePanda) [inline ouroboros-consensus/docs/Onboarding.md:149]: IMHO removed this file for now (in this PR). If we gonna need it (see my other PR) we will add it.

**Review by EncodePanda — CHANGES_REQUESTED** [review]
I love the level of detail @nfrisby You make all new-comers life much easier :)
Some nit-picking, few important bits.

File: ouroboros-consensus/docs/Contributing.md (hunk 2 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Contributing.md:5]: The concrete rules that come to mind are code formatting and the git process. Is that all of what you're referring to here? Am overlooking something else? I want to make sure we're talking about the same thing.

My current thought is that I don't see an alternative for this kind of project. If our plan were move fast, fail fast, iterate quickly to discover customer preferences etc, and so on, then it'd be different. Instead, the keyword I picked up on from transition discussions with Thomas and Edsko along these lines is _we're deliberate_. A structured approach seems appropriate for that.

My motivation for this PR (and the recent similar ones) is that, if we have rules, we should write them down. That helps both PR authors and reviewers approach the topic.

But I agree that the enumerated rules seem intimidating/overwhelming when collated like this. I think one actionable thing your comment here has prompted me to think about is explicitly explaining our expectations around enforcement of these rules. Basically "this is a list of things we will be looking for when we review PRs", "we don't expect PRs to start perfect", "you'll eventually internalize these", etc, etc.

Does that sound like a reasonable why to address the sentiment you shared in your comment here? Of course we can chat about this on our next call if you'd prefer.

File: ouroboros-consensus/docs/GitProcess.md (hunk 8 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/GitProcess.md:37]: That's a good article, yeah. In my experience, if it's legible and the content is taken seriously, that's enough. I think that's basically two of the seven rules in that article.

But obviously I'm not against codifying practices as rules ;). We just haven't prioritized them for commit messages yet.

File: ouroboros-consensus/docs/Contributing.md (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Contributing.md:11]: OK. I'll yield to your more recent onboarding experience. Mine was focused on testing, so this wasn't obvious to me; I can believe that was not the normal. 👍

File: ouroboros-consensus/docs/Contributing.md (hunk 7 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Contributing.md:33]: This is Markdown syntax declaring a synonym for a URL. If you view this file rendered (https://github.com/input-output-hk/ouroboros-network/blob/c65cf7a4412d5a84a0a03a50901310f79176eba5/ouroboros-consensus/docs/Contributing.md) you'll see that these lines are not themselves rendered.

File: ouroboros-consensus/docs/GitProcess.md (hunk 8 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/GitProcess.md:37]: I opened Issue #2914.

File: ouroboros-consensus/docs/GitProcess.md (hunk 10 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/GitProcess.md:74]: Good catch.

File: ouroboros-consensus/docs/GitProcess.md (hunk 11 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/GitProcess.md:79]: Good catch, I'll clarify.

File: ouroboros-consensus/docs/Onboarding.md (hunk 12 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Onboarding.md]: Let's leave it in until another PR rearranges/replaces that content. I like that it collates those all of those links and does so alongside a very high-level summary.

File: ouroboros-consensus/docs/GitProcess.md (hunk 9 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/GitProcess.md:26]: I have never asked. @dcoutts Was there specific logic to this decision?

My personal experience is that merge commits can themselves be useful, even if only for their timestamps and as an automated record of the PR number and description. Especially when we're making a reasonable effort to minimize the parallelism of the underlying genuine commits.

File: ouroboros-consensus/docs/Contributing.md (hunk 6 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Contributing.md]: Sounds good. Done.

File: ouroboros-consensus/docs/Contributing.md (hunk 5 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Contributing.md]: Makes sense. Done.

File: ouroboros-consensus/docs/Contributing.md (hunk 4 above)
Comment (nfrisby) [reply ouroboros-consensus/docs/Contributing.md]: Great!

**Comment by nfrisby** [comment]
Thanks for your review @EncodePanda ! I've Resolved the ones that I'm confident I addressed. Please feel free to unresolve them if you want to check and end up disagreeing.

File: ouroboros-consensus/docs/Contributing.md (hunk 2 above)
Comment (EncodePanda) [reply ouroboros-consensus/docs/Contributing.md]: > the keyword I picked up on from transition discussions with Thomas and Edsko along these lines is we're deliberate. A structured approach seems appropriate for that.

Yeah, I understand the intent, I was just referring to burned of someone who would like to do a small PR to our codebase but would be stopped by "the process".

> I think one actionable thing your comment here has prompted me to think about is explicitly explaining our expectations around enforcement of these rules. Basically "this is a list of things we will be looking for when we review PRs", "we don't expect PRs to start perfect", "you'll eventually internalize these", etc, etc.

👍  totally agree with that one, we can put that upfront in the `Contributing.md`. I think it does the issue directly, we keep our structured approach but we don't scare of newcommers.

File: ouroboros-consensus/docs/Onboarding.md (hunk 12 above)
Comment (EncodePanda) [reply ouroboros-consensus/docs/Onboarding.md]: 👍

**Review by EncodePanda — APPROVED** [review]

**Comment by nfrisby** [comment]
Thanks @EncodePanda . I added the sentence you suggested in `Contributing.md`, in addition to the small paragraph in each of the underlying guide documents 👍

**Pull request merged**